src/akar/util.clj

```diff
@@ -0,0 +1,42 @@
+(ns akar.util
+  (:require [clojure.walk :as cw]))
+
+(defn postwalk-fn-replace
+  "Calls the matching-fn on every expression and if matching-fn
+  returned true, replaces the expression by the result of calling
+  replacing-fn with the given expr as the parameter"
+  [matching-fn replacing-fn form]
+  (cw/postwalk (fn [x]
+                   (if (matching-fn x)
+                     (replacing-fn x)
+                     x))
+                 form))
+
+
+(defmacro def-trampolined-recur
+  "For recursive functions using recur along with akar match clauses,
+  this macro generates a tail call optimized function body using
+  trampoline.
+
+  This generates two functions,
+  1. Function with \"*\" suffixed to the name give. The body for
+  this function is same as the body given except the recur call replaced
+  with an anonymous function calling the same function recursively.
+
+  2. A function with the name - same as the name given in parameter. This
+  function calls the above generated function using trampoline.
+
+  This is done to avail the tail call optimization along with using the akar
+  patterns"
+  [nm args & body]
+  (let [tail-rec-fn-name (symbol (str nm "*"))
+        fn-body# (postwalk-fn-replace (fn [expr]
+                                        (and (seq? expr)
+                                             (= (first expr) 'recur)))
```

Comment:
Let's look for a distinct token than the built-in `recur` here; say `trampolined-recur`. In order to avoid throwing editor into a frenzy, we could define it as an unassigned var in the same namespace.

Reply:
Here is an example of a problem you could run into if you use `recur`:

Imagine someone is using our new `defn-trampolined` construct. They are making a call to some other macro, that expands inline to use `loop` and `recur`. Now `recur` his perfectly fine as it binds to `loop`. But our macro will rewrite it in a manner that it will no longer make sense.

Reply:
done